mixer/pkg/direct/envoymetrics/server.go

```diff
@@ -0,0 +1,386 @@
+	summaryCount    = "io.prometheus.client.metric.summary.sample_count"
+	summarySum      = "io.prometheus.client.metric.summary.sample_sum"
+	quantiles       = "io.prometheus.client.metric.summary.quantiles"
+	untypedVal      = "io.prometheus.client.metric.untyped.value"
+	histogramCount  = "io.prometheus.client.metric.histogram.sample_count"
+	histogramSum    = "io.prometheus.client.metric.histogram.sample_sum"
+	buckets         = "io.prometheus.client.metric.histogram.buckets"
+	metricTimestamp = "io.prometheus.client.metric.timestamp"
+)
+
+type (
+	attributes map[string]interface{}
+
+	server struct {
+		attrCh   chan attributes
+		reporter reporter
+	}
+
+	reporter interface {
+		Report([]attributes)
+	}
+
+	mixerReporter struct {
+		addr   string
+		client mixerpb.MixerClient
+	}
+
+	loggingReporter struct{}
+)
+
+var (
+	scope = log.RegisterScope("grpc", "grpc API server messages.", 0)
+
+	baseWords = []string{contextProtocol, nodeID, cluster, metadata, region, zone, subzone, buildVersion, familyName, familyType, metricLabels,
+		counterVal, gaugeVal, summaryCount, summarySum, quantiles, untypedVal, histogramCount, histogramSum, buckets, metricTimestamp}
+
+	attrBatchPool sync.Pool
+)
+
+// NewMetricsServiceServer returns a new server for the envoy v2 metrics service.
+func NewMetricsServiceServer(mixerAddr string, batchSize int64) (mpb.MetricsServiceServer, error) {
+	attrBatchPool = sync.Pool{
+		New: func() interface{} {
+			return make([]attributes, 0, batchSize)
+		},
+	}
+	ch := make(chan attributes, batchSize)
+
+	var r reporter
+	if len(mixerAddr) > 0 {
+		scope.Warnf("Configuring Mixer reporter with address: %v", mixerAddr)
+		client, err := mixerAPIClient(mixerAddr)
+		if err != nil {
+			return nil, fmt.Errorf("could not build mixer client: %v", err)
+		}
+		r = &mixerReporter{addr: mixerAddr, client: client}
+	} else {
+		r = &loggingReporter{}
+	}
+
+	go processAttributes(ch, batchSize, r)
+	return &server{attrCh: ch}, nil
+}
+
+func (s *server) StreamMetrics(stream mpb.MetricsService_StreamMetricsServer) error {
+	scope.Debug("new metrics service stream established")
+	var idAttrs attributes
+	for {
+		msg, err := stream.Recv()
+		if err != nil {
+			if status.Code(err) != codes.Canceled {
+				scope.Errorf("error in processing stream: %v", err)
+			}
+			return err
+		}
+		if msg.Identifier != nil {
+			idAttrs = identiferAttributes(msg.Identifier)
+			scope.Debugf("id attrs for stream: %#v", idAttrs)
+		}
+		for _, metricsFam := range msg.EnvoyMetrics {
+			for _, m := range metricsFam.Metric {
+				attrs := metricsAttributes(metricsFam, m)
+				s.attrCh <- merge(idAttrs, attrs)
+			}
+		}
+	}
+}
+
+func identiferAttributes(identifier *mpb.StreamMetricsMessage_Identifier) attributes {
+	attrs := make(attributes, 7)
+
+	node := identifier.GetNode()
+	// TODO: figure out how best to serialize metadata
+	if c := node.GetCluster(); len(c) > 0 {
+		attrs[cluster] = c
+	}
```

Comment:
You cannot get the local cluster of a node. We don't set it.
We do set the labels for the node. If you want us to set any other metadata for a node, let us know.